sdcflows/transform.py

```diff
@@ -72,151 +183,189 @@ def fit(self, spatialimage):
     ):
         """
         Apply a transformation to an image, resampling on the reference spatial object.
 
+        Handles parallelization to resample 4D images.
+
         Parameters
         ----------
-        spatialimage : `spatialimage`
+        moving : :obj:`~nibabel.spatialimages.SpatialImage`
             The image object containing the data to be resampled in reference
             space
-        reference : spatial object, optional
-            The image, surface, or combination thereof containing the coordinates
-            of samples that will be sampled.
-        order : int, optional
+        pe_dir : :obj:`str`
+            A valid ``PhaseEncodingDirection`` metadata value.
+        ro_time : :obj:`float`
+            The total readout time in seconds.
+        xfms : :obj:`None` or :obj:`list`
+            A list of rigid-body transformations previously estimated that will
+            realign the dataset (that is, compensate for head motion) after resampling.
+        order : :obj:`int`, optional
             The order of the spline interpolation, default is 3.
             The order has to be in the range 0-5.
         mode : {'constant', 'reflect', 'nearest', 'mirror', 'wrap'}, optional
             Determines how the input image is extended when the resamplings overflows
             a border. Default is 'constant'.
         cval : float, optional
             Constant value for ``mode='constant'``. Default is 0.0.
-        prefilter: bool, optional
+        prefilter : :obj:`bool`, optional
             Determines if the image's data array is prefiltered with
             a spline filter before interpolation. The default is ``True``,
             which will create a temporary *float64* array of filtered values
             if *order > 1*. If setting this to ``False``, the output will be
             slightly blurred if *order > 1*, unless the input is prefiltered,
             i.e. it is the result of calling the spline filter on the original
             input.
+        output_dtype : :obj:`str` or :obj:`~numpy.dtype`
+            Override the output data type, instead of propagating it from the
+            moving image.
 
         Returns
         -------
-        resampled : `spatialimage` or ndarray
+        resampled : :obj:`~nibabel.spatialimages.SpatialImage`
             The data imaged after resampling to reference space.
 
         """
         from sdcflows.utils.tools import ensure_positive_cosines
 
         # Ensure the fmap has been computed
-        if isinstance(spatialimage, (str, bytes, Path)):
-            spatialimage = nb.load(spatialimage)
-
-        spatialimage, axcodes = ensure_positive_cosines(spatialimage)
+        if isinstance(moving, (str, bytes, Path)):
+            moving = nb.load(moving)
 
-        self.fit(spatialimage)
-        fmap = self.mapped.get_fdata().copy()
+        # Generate warp field (before ensuring positive cosines)
+        self.fit(moving)
```

Comment:
I think this undoes the `unwarp.fit()` done in bspline.py?

Reply:
It shouldn't - the idea was that it gets cached in the first run, and when the moving has the same characteristics (same matrix and affine) then you can go ahead (even if you actually changed the data because it is matched and you know what you are doing).

This second fit is to make sure we do not apply before checking the moving and fitting the bsplines if necessary.

However, I agree there could be some error in the logic because in my testing it did look like it was undoing the previous fit. I thought it was because I was ensuring positive cosines first. I have it as something I need to double check before merge.

Reply:
I think we want to separate fit from validate. Apply should raise an exception on mismatch, not silently drop the important bits.

Reply:
I'm going through the code and I totally agree. Validation should be separate